scipy/optimize/_dual_annealing.py

```diff
@@ -0,0 +1,631 @@
+        # Visiting distribution instance
+        self.visit_dist = visit_dist
+        # Wrapper to objective function
+        self.func_wrapper = func_wrapper
+        # Wrapper to the local minimizer
+        self.minimizer_wrapper = minimizer_wrapper
+        self.callback = callback
+        self.not_improved_idx = 0
+        self.not_improved_max_idx = 1000
+        self._rand_state = rand_state
+        self.temperature_step = 0
+        self.K = 100 * len(energy_state.current_location)
+
+    def accept_reject(self, j, e, x_visit):
+        r = self._rand_state.random_sample()
+        pqv_temp = (self.acceptance_param - 1.0) * (
+            e - self.energy_state.current_energy) / (
+                self.temperature_step + 1.)
+        if pqv_temp <= 0.:
+            pqv = 0.
+        else:
+            pqv = np.exp(np.log(pqv_temp) / (
+                1. - self.acceptance_param))
+        if r <= pqv:
+            # We accept the new location and update state
+            self.energy_state.current_energy = e
+            self.energy_state.current_location = np.copy(x_visit)
+            self.xmin = np.copy(self.energy_state.current_location)
+
+        # No improvement for a long time
+        if self.not_improved_idx >= self.not_improved_max_idx:
+            if not j or self.energy_state.current_energy < self.emin:
+                self.emin = self.energy_state.current_energy
+                self.xmin = np.copy(self.energy_state.current_location)
+
+    def run(self, step, temperature):
+        self.temperature_step = temperature / float(step + 1)
+        self.not_improved_idx += 1
+        for j in range(self.energy_state.current_location.size * 2):
+            if not j:
+                if not step:
+                    self.energy_state_improved = True
+                else:
+                    self.energy_state_improved = False
+            x_visit = self.visit_dist.visiting(
+                self.energy_state.current_location, j, temperature)
+            # Calling the objective function
+            e = self.func_wrapper.fun(x_visit)
+            if e < self.energy_state.current_energy:
+                # We have got a better energy value
+                self.energy_state.current_energy = e
+                self.energy_state.current_location = np.copy(x_visit)
+                if e < self.energy_state.ebest:
+                    val = self.track_better_minimum(e, x_visit, 0)
+                    if not val:
+                        return False
+                    self.energy_state_improved = True
+                    self.not_improved_idx = 0
+            else:
+                # We have not improved but do we accept the new location?
+                self.accept_reject(j, e, x_visit)
+            if self.func_wrapper.nfev >= self.func_wrapper.maxfun:
+                return False
+        # End of MarkovChain loop
+        return True
+
+    def track_better_minimum(self, e, x, context):
```

Comment:
Suggestion: I think this function fits better as part of EnergyState.  Then you could wrap this functionality into an update method similar to basinhopping.Storage.update().  Then EnergyState has all the responsibility for keeping track of the lowest state and MonteCarlo doesn't have to worry about it anymore.

Reply:
Agreed. Moved to EnergyState class.

Reply:
Added methods update_current and update_best (which also trigger the callback if defined) in EnergyState to address that. Thanks!